cmd/zetaclientd/start.go

```diff
@@ -210,6 +213,12 @@ func start(_ *cobra.Command, _ []string) error {
 	// Set P2P ID for telemetry
 	telemetryServer.SetP2PID(server.GetLocalPeerID())
 
+	// Create a notification context for background threads.These threads are responsible for sending shutdown signals to the main thread.
+	notifyCtx, cancel := context.WithCancelCause(ctx)
```

Comment:
This could be refined so we reduce the codebase:
```suggestion
	ctxwc, cancel := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```

So we can remove all the code regarding the sig channel, saving resources and logic.

And replace this:
```go
	startLogger.Info().Msgf("awaiting shutdown signals")
	select {
	case <-notifyCtx.Done():
		cause := context.Cause(notifyCtx)
		startLogger.Info().Msgf("shutdown signal received , cause : %s", cause)

	case sig := <-ch:
		startLogger.Info().Msgf("stop signal received: %s", sig)
	}

	//stop chain observers
	for _, observer := range observerMap {
		observer.Stop()
	}
	zetacoreClient.Stop()

	return nil
```

for this leaner code:
```go
	startLogger.Info().Msgf("zetaclientd is running")
	<-notifyCtx.Done()

        startLogger.Info().Msgf("zetaclientd shutting down")
	for _, observer := range observerMap {
		observer.Stop()
	}

	zetacoreClient.Stop()

	return nil
```

Reply:
I think , we would have to define a custom error , to fetch the cause in this case .
I want to use the cause to print out which job sends the shutdown request . 
What do you think ?

Reply:
In the case of an error that would be printed anyway, right? as we execute the `bg.Work` routines with `WithLogger` and `WithName` the first error log we have before "zetaclientd shutting down" we can assume it's the one calling `cancel()`.

Another way to do it, more deterministic, is to modify the context that we create in StartTssMigrationRoutines and return it along with the cancelCauseFunc:
```go
	causeCtx, causeCancel := context.WithCancelCause(ctx)
	bg.Work(
		migrationRoutinesContext,
		c.HandleTSSUpdate,
		bg.WithName("HandleTSSUpdate"),
		bg.WithLogger(masterLogger),
		bg.WithCallback(causeCancel),
...

return migrationRoutinesContext, cancel
```
When the first task fails will cancel causeCtx:
- all the other tasks should receive a ctx.Done() and close.
- the `causeCtx` would carry the error and we can recover it from the main goroutine.

This would require some prior testing, as I'm just talking from memory and haven't tested it myself.

Reply:
I think your first point makes sense , we are printing inside the `Work` function , I can remove the cancelCause